object/slicer/slicer.go

```diff
@@ -316,18 +340,65 @@ func (x *PayloadWriter) Write(chunk []byte) (int, error) {
 		return 0, nil
 	}
 
-	alreadyWritten := x.rootMeta.length
+	buffered := x.rootMeta.length
 	if x.withSplit {
-		alreadyWritten = x.childMeta.length
+		buffered = x.childMeta.length
 	}
 
-	if alreadyWritten+uint64(len(chunk)) <= x.payloadSizeLimit {
-		return x.currentWriter.Write(chunk)
+	if buffered+uint64(len(chunk)) <= x.payloadSizeLimit {
+		// buffer data to produce as few objects as possible for better storage efficiency
+		_, err := x.metaWriter.Write(chunk)
+		if err != nil {
+			return 0, err
+		}
+
+		var n int
+		payloadBufferLen := uint64(len(x.payloadBuffer))
+		if payloadBufferLen > buffered {
+			n = copy(x.payloadBuffer[buffered:], chunk)
+			if n == len(chunk) {
+				return n, nil
+			}
+
+			chunk = chunk[n:]
+			buffered += uint64(n)
+		}
+
+		if x.extraPayloadBuffer == nil {
+			// if here for the first time, then allocate the minimum buffer sufficient for
+			// writing: user may do one Write followed by Close. In such cases there is no
+			// point in allocating a buffer of payloadSizeLimit size.
+			// TODO(#544): support external buffer pools
+			x.extraPayloadBuffer = make([]byte, len(chunk))
+		} else if payloadBufferLen+uint64(len(x.extraPayloadBuffer)) < x.payloadSizeLimit {
+			x.extraPayloadBuffer = append(x.extraPayloadBuffer, make([]byte, x.payloadSizeLimit-buffered)...)
```

Comment:
have you checked if it is ok to create a new slice of zeros to append it? isn't `make([]byte, len(extraPayloadBuffer)+x.payloadSizeLimit-buffered)` and `copy` more optimal?

Reply:
actually i didnt thought about efficiency when wrote this instruction. Quick benchmarks say they are almost the same

so more readable `make+copy` wins